src/env.rs

```diff
@@ -0,0 +1,233 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An environment which is passed downward into the widget tree.
+
+use std::collections::HashMap;
+use std::convert::{TryFrom, TryInto};
+use std::fmt::{Debug, Formatter};
+use std::marker::PhantomData;
+use std::ops::Deref;
+use std::sync::Arc;
+
+use crate::kurbo::{Point, Rect, Size};
+use crate::piet::Color;
+
+use crate::Data;
+
+/// An environment passed down through all widget traversals.
+///
+/// All widget methods have access to an environment, and it is passed
+/// downwards during traversals.
+///
+/// A widget can retrieve theme parameters (colors, dimensions, etc.). In
+/// addition, it can pass custom data down to all descendants. An important
+/// example of the latter is setting a value for enabled/disabled status
+/// so that an entire subtree can be disabled ("grayed out") with one
+/// setting.
+#[derive(Clone, Default)]
+pub struct Env(Arc<EnvImpl>);
+
+#[derive(Clone, Default)]
+struct EnvImpl {
+    map: HashMap<String, Value>,
+}
+
+/// A typed key.
+///
+/// This lets you retrieve values of a given type.
+pub struct Key<T> {
+    key: &'static str,
+    value_type: PhantomData<T>,
+}
+
+// we could do some serious deriving here: the set of types that can be stored
+// could be defined per-app
+// Also consider Box<Any> (though this would also impact debug).
+#[derive(Clone)]
+pub enum Value {
+    Point(Point),
+    Size(Size),
+    Rect(Rect),
+    Color(Color),
+    Float(f64),
+    String(String),
+}
+
+impl Env {
+    // TODO: want to change this to return `&V`.
+    pub fn get<V: TryFrom<Value, Error = String>>(&self, key: Key<V>) -> V {
+        if let Some(value) = self.0.map.get(key.key) {
+            value.to_owned().into_inner_unchecked()
+        } else {
+            panic!("key for {} not found", key.key)
+        }
+    }
+
+    // Also &V.
+    pub fn try_get<V: TryFrom<Value, Error = String>>(&self, key: Key<V>) -> Option<V> {
+        self.0
+            .map
+            .get(key.key)
+            .map(|value| value.to_owned().into_inner_unchecked())
+    }
+
+    /// Adds a key/value, acting like a builder.
+    pub fn adding<K: Into<String>, V: Into<Value>>(mut self, key: K, value: V) -> Env {
```

Comment:
I want to highlight this as something that is a bit unidiomatic; it's basically bundling a builder pattern into the type itself. I like this because you don't need to import a separate builder type and you also don't need to call `build`, but it might seem a bit weird to some folks.